### PR TITLE
IGNITE-14413 Added start of new node from cli interface.

### PR DESCRIPTION
**Pull request by kgusakov**

### CODE REVIEW

File: modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java [hunk 1]
```diff
@@ -171,6 +175,31 @@ public boolean bootstrapped() {
...
+    }
+
+    /**
+     * @return Node name, if was stored earlier.
+     * @throws IgniteInternalCheckedException If couldn't get node name from the vault.
+     */
+    public String name() throws IgniteInternalCheckedException {
+        synchronized (mux) {
+            try {
+                return new String(vaultService.get(NODE_NAME).get().value());
```
Comment (alievmirza) [inline modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java:195]: Please specify charset StandardCharsets.UTF_8

Comment (alievmirza) [inline modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java:195]: I would wait for future's get outside of the synchronized block

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgnitionImpl.java [hunk 2]
```diff
@@ -80,11 +80,14 @@
     private static final String VER_KEY = "version";
 
     /** {@inheritDoc} */
-    @Override public synchronized Ignite start(String jsonStrBootstrapCfg) {
+    @Override public synchronized Ignite start(String nodeName, String jsonStrBootstrapCfg) {
+        assert !StringUtil.isNullOrEmpty(nodeName) : "Node local name is empty";
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgnitionImpl.java:84]: I would add @NotNull for nodeName here and in ancestor, and also reflect that facts in javadocs

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                    throw new ParseException(usage);
+            }
+            else
+                throw new ParseException(usage);
+        }
+
+        /**
+         * Exception for indicate any problems with parsing CLI args.
+         */
+        private static class ParseException extends Exception {
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:102]: Should it extends some Ignite related exception?

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+         * @param nodeName Name of the node.
+         * @param config Path to config file.
+         */
+        private Args(String nodeName, Path config) {
+            this.nodeName = nodeName;
+            this.config = config;
+        }
+
+        /**
+         * Simple CLI arguments parser.
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:80]: Should we reflect in javadoc here or somewhere else, how many arguments we are waiting for and which arguments?

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+         * @param args CLI arguments.
+         * @return Parsed arguments.
+         * @throws ParseException if required args are absent.
+         */
+        private static Args parseArgs(String[] args) throws ParseException {
+            if (args.length == 1)
+                return new Args(args[0], null);
+            else if (args.length == 3) {
+                if ("--config".equals(args[0]))
+                    return new Args(args[2], Path.of(args[1]));
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:91]: Should there be some validation? `Path.of` could throw  at least `IllegalArgumentException` or `FileSystemNotFoundException`, should we catch such cases and rethrow `ParseException`?

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        try {
+            parsedArgs = Args.parseArgs(args);
+        }
+        catch (Args.ParseException e) {
+            System.out.println(e.getMessage());
+            System.exit(1);
+        }
+
+        String jsonCfgStr = null;
+        if (parsedArgs.config != null)
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:47]: empty line

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    public static void main(String[] args) throws IOException {
+        Args parsedArgs = null;
+
+        try {
+            parsedArgs = Args.parseArgs(args);
+        }
+        catch (Args.ParseException e) {
+            System.out.println(e.getMessage());
+            System.exit(1);
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:43]: empty line. Also, should we write here some clarification what happened before `e.getMessage()`?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/IgnitionTest.java [hunk 8]
```diff
@@ -87,7 +85,7 @@ void testNodesStartWithBootstrapConfiguration() {
     @Test
     @Disabled("https://issues.apache.org/jira/browse/IGNITE-14709")
     void testNodeStartWithoutBootstrapConfiguration() {
-        Ignite ignite = IgnitionManager.start(null);
+        Ignite ignite = IgnitionManager.start(null, null);
```
Comment (alievmirza) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/runner/app/IgnitionTest.java:88]: what we expect when we start node without bootstarping configuration and with null as name? As far as  I can understand, we'll fall with assert, looks strange

File: modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java (hunk 1 above)
Comment (alievmirza) [inline modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java:195]: I'm not sure, but should we consider some caching here? Seems that `name()` might be called frequently, WDYT?

**Review by alievmirza — CHANGES_REQUESTED** [review]

File: modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java (hunk 1 above)
Comment (kgusakov) [reply modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java]: Looks like this pattern can break some invariants, which were the causes for mutex in general. As we discussed earlier - this pattern of waiting is used in all methods of VaultManager.

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java (hunk 5 above)
Comment (kgusakov) [reply modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java]: Fixed for `InvalidPathException`. Looks like other types of runtime exceptions are not relevant for catching here.

File: modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java (hunk 7 above)
Comment (kgusakov) [reply modules/runner/src/main/java/org/apache/ignite/app/IgniteCliRunner.java:54]: It's so tricky to improve the messaging for all types of issues without more complex argument parsing logic. I think that usage message + [some details] (for --config path parsing) is ok for current arguments. WDYT?

Anyway, error reporting updated, you can check the current aprroach.

File: modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java (hunk 1 above)
Comment (kgusakov) [reply modules/vault/src/main/java/org/apache/ignite/internal/vault/VaultManager.java]: Maybe, not sure too :) `appliedRevision` can have the same issue, maybe we should fix it in the scope of another issue?

File: modules/runner/src/main/java/org/apache/ignite/internal/app/IgnitionImpl.java [hunk 9]
```diff
@@ -80,11 +82,14 @@
...
     /** {@inheritDoc} */
-    @Override public synchronized Ignite start(String jsonStrBootstrapCfg) {
+    @Override public synchronized Ignite start(@NotNull String nodeName, @Nullable String jsonStrBootstrapCfg) {
+        assert !StringUtil.isNullOrEmpty(nodeName) : "Node local name is empty";
+
         ackBanner();
 
         // Vault Component startup.
         VaultManager vaultMgr = new VaultManager(new VaultServiceImpl());
+        vaultMgr.putName(nodeName).join();
```
Comment (alievmirza) [inline modules/runner/src/main/java/org/apache/ignite/internal/app/IgnitionImpl.java:92]: empty line

**Pull request closed without merging**